modules/acm_sku/src/ProductOptionsManager.php

```diff
@@ -202,6 +208,55 @@ public function synchronizeProductOptions() {
       foreach ($option_sets as $options) {
         foreach ($options['options'] as $key => $value) {
           $this->createProductOption($langcode, $key, $value, $options['attribute_id'], $options['attribute_code'], $weight++);
+          $options_available[$options['attribute_code']][$options['attribute_id']] = $options['attribute_id'];
+        }
+      }
+    }
+
+    if ($options_available) {
+      $this->deleteUnavailableOptions($options_available);
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function deleteUnavailableOptions(array $synced_options) {
+    // Cleanup queries can be done only for one language.
+    $query = $this->connection->select('taxonomy_term__field_sku_attribute_code', 'ttfsac');
+    $query->addExpression('count(entity_id)', 'cnt');
+    $query->condition('field_sku_attribute_code_value', array_keys($synced_options), 'IN');
+    $result = $query->execute()->fetchAllKeyed(0, 0);
+    $options_in_db = reset($result);
+
+    $synced_option_ids = [];
+    foreach ($synced_options as $attribute_code => $options) {
+      $synced_option_ids = array_merge($synced_option_ids, $options);
+    }
+
+    // Do nothing if count of option ids synced and in DB match.
+    if (count($synced_option_ids) === $options_in_db) {
+      return;
+    }
+
+    $query = $this->termStorage->getQuery();
+    $query->condition('field_sku_option_id', $synced_option_ids, 'NOT IN');
+    $query->condition('field_sku_attribute_code', array_keys($synced_options), 'IN');
+    $query->condition('vid', self::PRODUCT_OPTIONS_VOCABULARY);
+    $tids = $query->execute();
+
+    if ($tids) {
+      foreach (array_chunk($tids, 50) as $ids) {
```

Comment:
is this for memory consumption optimisation?

Reply:
yes, to avoid loading large number of entities

cc @joshirohit100

Reply:
yes! this is done to handle memory issue as below we doing -
```
$entities = $this->termStorage->loadMultiple($ids)
$this->termStorage->delete($entities);
```